### ISSUE TITLE
GME files not recognized (French pen)

### ISSUE THREAD

**Issue by severin-lemaignan** [entropia/tip-toi-reveng#143]
Hello,

I'm struggling getting my tiptoi to accept `tttool` generated GMEs.

For instance, I've tried to assemble `Debug.yaml` (after setting the product id to 19 to match the value returned by `tttool info Livre\ foret.gme`, one of the books we have here). `tttool` assembles it fine, but when  copied on the pen, none of the book's OIDs is recognized (the pen keeps complaining that I should first download the audio files corresponding to my product).

To make sure, I've set as well `gme-lang: FRENCH`, to no difference.

I'm using a French pen, updated to the firmware 3202.

Any idea?

Issue initially reported here thawn/ttmp32gme#5

**Comment by nomeata** [comment]
The support for `gme-lang` is rather new, and not part of any released version. Are you using the repository version? It migt also not be perfect here; if you can find out what needs to change about the GME file to fix this, let me know!

**Comment by severin-lemaignan** [comment]
I'm using the repo, yes.
The language tag look fine -- at least, when read again with `tttool info`, it looks identical to the original book. To be fair, I do not even know if the language needs to be set. I only tried because it was mentioned as smthg required in #132

**Comment by nomeata** [comment]
> at least, when read again with tttool info

That doesn’t say much. If we misunderstand the fileformat, then `tttool info` will do the same mistake.

**Comment by uli42** [comment]
On Sat, Feb 11, 2017 at 10:34 PM, Séverin Lemaignan
<notifications@github.com> wrote:
>
> Hello,
>
> I'm struggling getting my tiptoi to accept tttool generated GMEs.
>
> For instance, I've tried to assemble Debug.yaml (after setting the product id to 19 to match the value returned by tttool info Livre\ foret.gme, one of the books we have here). tttool assembles it fine, but when copied on the pen, none of the book's OIDs is recognized (the pen keeps complaining that I should first download the audio files corresponding to my product).
>
> To make sure, I've set as well gme-lang: FRENCH, to no difference.
>
> I'm using a French pen, updated to the firmware 3202.
>
> Any idea?

Is this an old or new (with musicplayer) pen? For the old ones (maybe
also for the new ones, but AFAIK this is untested) you can find a
description how to switch the language of the pen on the "Firmware"
page in the wiki. You could check if it works for a german pen. If so,
we  have not fully understood the file format, it seems.

Uli

**Comment by severin-lemaignan** [comment]
@uli42 hum, interesting. How can I distinguish between an old and new pen ? (I'm not sure about what the music player is supposed to be)

I'll give a try to reset the language.

**Comment by severin-lemaignan** [comment]
Alright, so:
- I've switched back to German: As only `Update3202.upd` is available from Ravensburger, I've had to switch from `Update3202MT_fr.upd` in French to `Update3202.upd` in German -- ie, without the `MT`. I did so by changing the firmware name in `LanguageInfoMT.txt`.
- after a reboot, the pen successfully update the firmware (in German), and my `Debug.gme` worked. Cool.
- I decided to switch back to French:
  - As I was now using the firmware without the `MT`, I first had to create a `LanguageInfo.txt`. I did so by copy-pasting my `LanguageInfoMT.txt` (and re-swapping back the language codes)
  - when rebooting, the pen started the firmware update (in French), but... it *never finished!* and the pen looks bricked now. Can not stop/start it. If I plug it over USB, it is not seen by the computer. I've tried to remove the battery. No difference :-(( My children are going to be upset if I can not find a way to get the pen working again! :-/

**Comment by uli42** [comment]
On Sun, Feb 12, 2017 at 10:45 PM, Séverin Lemaignan
<notifications@github.com> wrote:
>
> Alright, so:
>
> I've switched back to German: As only Update3202.upd is available from Ravensburger, I've had to switch from Update3202MT_fr.upd in French to Update3202.upd in German -- ie, without the MT. I did so by changing the firmware name in LanguageInfoMT.txt.
> after a reboot, the pen successfully update the firmware (in German), and my Debug.gme worked. Cool.

So we have to find out what's different on french pens.

According to the firmware page
https://github.com/entropia/tip-toi-reveng/wiki/Firmware a German _MT
version exists. And I have it on my harddisk (the numbers are md5
checksums):

ca6696b0f57672d90e509c39a2e62356  update3202MT.upd
683e4d0a07f71e0151106c889b48d498  Update3202.upd

I also both versions in French are existing, too:

36edfdfe07d7d1e864d40f953f2d401b  Update3202_fr.upd
3788fc92ad5b3d8d4691571ebb9be405  Update3202MT_fr.upd

> I decided to switch back to French:

> As I was now using the firmware without the MT, I first had to create a LanguageInfo.txt. I did so by copy-pasting my LanguageInfoMT.txt (and re-swapping back the language codes)
> when rebooting, the pen started the firmware update (in French), but... it never finished! and the pen looks bricked now. Can not stop/start it. If I plug it over USB, it is not seen by the computer. I've tried to remove the battery. No difference :-(( My children are going to be upset if I can not find a way to get the pen working again! :-/

Ohoh, that was risky. As he wiki states:

It is currently unknown what MT in some of the file names stands for
(maybe there are different hardware revisions that need a slightly
different firmware; current suspicion is that the MT firmware supports
Micron Technology flash memory for the 4 GB version of the pen).

What size does your pen have?

On the same wiki page you can find how to enter the test mode. Does
that still work?

It could be that your pen shows up with a difference USB id and your
OS (Wndows?) does not recognize. Please read the section about USB
Storage Boot Mode, maybe your pen is in that mode now (which wouldn't
help in your situation, but will eventually provide further insight).

Uli

**Comment by uli42** [comment]
One more thing: the URLs for the German firmwares (from
http://static.tiptoi.com/db/tiptoi_de_de.xml)

  <firmware>
    <firmware1>
      <version>136</version>
      <path>http://static.tiptoi.com/db/firmware/update encrypt normal
freq.upd</path>
      <checksum>28F7854A</checksum>
    </firmware1>
    <firmware2>
      <version>38</version>
      <path>http://static.tiptoi.com/db/firmware/Update3202_dl.upd</path>
      <checksum>39EC05ED</checksum>
    </firmware2>
    <firmware3>
      <version>38</version>
      <path>http://static.tiptoi.com/db/firmware/update3202MT.upd</path>
      <checksum>3A04DD92</checksum>
    </firmware3>
  </firmware>

and for the French ones (from http://static.tiptoi.com/db/tiptoi_fr_fr.xml):

  <firmware>
    <firmware1>
      <version>136</version>
      <path>http://static.tiptoi.com/db/firmware/update encrypt normal
freq.upd</path>
      <checksum>28F7854A</checksum>
    </firmware1>
    <firmware2>
      <version>21</version>
      <path>http://static.tiptoi.com/db/firmware/Update3202_fr.upd</path>
      <checksum>35F285D8</checksum>
    </firmware2>
    <firmware3>
      <version>23</version>
      <path>http://static.tiptoi.com/db/firmware/Update3202MT_fr.upd</path>
      <checksum>360762FE</checksum>
    </firmware3>

**Comment by severin-lemaignan** [comment]
On 12/02/2017 22:11, Ulrich Sibiller wrote:
> On Sun, Feb 12, 2017 at 10:45 PM, Séverin Lemaignan
> <notifications@github.com> wrote:
>>
>> Alright, so:
>>
>> I've switched back to German: As only Update3202.upd is available from
> Ravensburger, I've had to switch from Update3202MT_fr.upd in French to
> Update3202.upd in German -- ie, without the MT. I did so by changing the
> firmware name in LanguageInfoMT.txt.
>> after a reboot, the pen successfully update the firmware (in German),
> and my Debug.gme worked. Cool.
>
> So we have to find out what's different on french pens.
>
> According to the firmware page
> https://github.com/entropia/tip-toi-reveng/wiki/Firmware a German _MT
> version exists. And I have it on my harddisk (the numbers are md5
> checksums):
>
> ca6696b0f57672d90e509c39a2e62356 update3202MT.upd
> 683e4d0a07f71e0151106c889b48d498 Update3202.upd

Yes, I've indeed read on the wiki that the German `MT` does exist, but I
could not immediately find it.

> I also both versions in French are existing, too:
>
> 36edfdfe07d7d1e864d40f953f2d401b Update3202_fr.upd
> 3788fc92ad5b3d8d4691571ebb9be405 Update3202MT_fr.upd

Yes, these two can be downloaded from ravensburger website.

>
>> I decided to switch back to French:
>
>> As I was now using the firmware without the MT, I first had to create
> a LanguageInfo.txt. I did so by copy-pasting my LanguageInfoMT.txt (and
> re-swapping back the language codes)
>> when rebooting, the pen started the firmware update (in French),
> but... it never finished! and the pen looks bricked now. Can not
> stop/start it. If I plug it over USB, it is not seen by the computer.
> I've tried to remove the battery. No difference :-(( My children are
> going to be upset if I can not find a way to get the pen working again! :-/
>
> Ohoh, that was risky. As he wiki states:
>
> It is currently unknown what MT in some of the file names stands for
> (maybe there are different hardware revisions that need a slightly
> different firmware; current suspicion is that the MT firmware supports
> Micron Technology flash memory for the 4 GB version of the pen).
>
> What size does your pen have?

4GB.
Yes, it was a bit risky :-)
If only that, going from a `MT` version to a non-`MT` one was successful.

> On the same wiki page you can find how to enter the test mode. Does
> that still work?

Nope :-/
[and I tested before, it did work]

> It could be that your pen shows up with a difference USB id and your
> OS (Wndows?) does not recognize. Please read the section about USB
> Storage Boot Mode, maybe your pen is in that mode now (which wouldn't
> help in your situation, but will eventually provide further insight).

I'm on Linux.
Plugging/unplugging shows nothing at all, neither with `dmesg` nor with
`lsusb`... So I'm quite pessimistic :-/
________________________________
[http://www.plymouth.ac.uk/images/email_footer.gif]<http://www.plymouth.ac.uk/worldclass>

This email and any files with it are confidential and intended solely for the use of the recipient to whom it is addressed. If you are not the intended recipient then copying, distribution or other use of the information contained is strictly prohibited and you should not rely on it. If you have received this email in error please let the sender know immediately and delete it from your system(s). Internet emails are not necessarily secure. While we take every care, Plymouth University accepts no responsibility for viruses and it is your responsibility to scan emails and their attachments. Plymouth University does not accept responsibility for any changes made after it was sent. Nothing in this email or its attachments constitutes an order for goods or services unless accompanied by an official order form.

**Comment by severin-lemaignan** [comment]
Ok, so by following the procedure on the `Firmware` page, I've been able to boot in USB Storage Boot Mode. Good news: the pen is not dead! :-)

Not sure what to do next, though!

**Comment by severin-lemaignan** [comment]
So, once in USB Storage Boot Mode, I've fired up my Windows 10 virtual machine, with Anyka's BurnTool 3.0.18.2. BurnTool sees the tiptoi and happily imports `Update3202MT_fr.upd`.

What to do next is less clear. I've ticked both `Update mode` and `UDisk Burn` and then clicked on `Download`. It apparently did something as after ~20 sec, the tiptoi eventually disconnected. But it did not reboot either...

**Comment by uli42** [comment]
On Mon, Feb 13, 2017 at 12:41 AM, Séverin Lemaignan
<notifications@github.com> wrote:
>
> So, once in USB Storage Boot Mode, I've fired up my Windows 10 virtual machine, with Anyka's BurnTool 3.0.18.2. BurnTool sees the tiptoi and happily imports Update3202MT_fr.upd.
>
> What to do next is less clear. I've ticked both Update mode and UDisk Burn and then clicked on Download. It apparently did something as after ~20 sec, the tiptoi eventually disconnected. But it did not reboot either...

That's the problem: unfortunately noone has managed yet to flash a
tiptoi with burntool. Burntool creates a logfil, maybe the errors in
there can help.

Uli

**Comment by thawn** [comment]
If I remember correctly, for the ITALIAN language version, the total length of the language tag seemed to be important. It may be worth a try (once the pen is hopefully fixed :) ) to make sure that the total length of the language tag in the GME file is the same as in other FRENCH gme files.

just my 2 cents

Till

**Comment by chrisz** [comment]
Hello,

I encountered the same problem with the french pen I bought last week.

I made different tests and it works if I had:
gme-lang: FRENCH
comment: CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316

But it does'nt work with:
- no gme-lang and no comment
- no gme-lang + comment "CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316"
- gme-lang FRENCH + comment "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" (size of the first comment - 1 character)

And it works with:
- gme-lang FRENCH + comment "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" (same size as the first comment)

So it seems the comment must have 48 characters and language is required for french pen.

Thanks for this great tool :-) !

Chris

**Comment by rafiberlin** [comment]
> Hello,
> 
> I encountered the same problem with the french pen I bought last week.
> 
> I made different tests and it works if I had:
> gme-lang: FRENCH
> comment: CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316
> 
> But it does'nt work with:
> 
> * no gme-lang and no comment
> * no gme-lang + comment "CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316"
> * gme-lang FRENCH + comment "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" (size of the first comment - 1 character)
> 
> And it works with:
> 
> * gme-lang FRENCH + comment "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" (same size as the first comment)
> 
> So it seems the comment must have 48 characters and language is required for french pen.
> 
> Thanks for this great tool :-) !
> 
> Chris

Hi Chris,

Could you describe how you were able to do that? (I am new to this tool and I am struggling with the French Firmware...)

Thanks a lot!

Rafi

**Comment by uli42** [comment]
On Tue, Sep 18, 2018 at 11:04 AM, rafiberlin <notifications@github.com> wrote:
>
> Hello,
>
> I encountered the same problem with the french pen I bought last week.
>
> I made different tests and it works if I had:
> gme-lang: FRENCH
> comment: CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316
>
> But it does'nt work with:
>
> no gme-lang and no comment
> no gme-lang + comment "CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316"
> gme-lang FRENCH + comment "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" (size of the first comment - 1 character)
>
> And it works with:
>
> gme-lang FRENCH + comment "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" (same size as the first comment)
>
> So it seems the comment must have 48 characters and language is required for french pen.


Well, the gme format also includes the length of that string, see
here: https://github.com/entropia/tip-toi-reveng/blob/master/GME-Format.md

So you must probably adjust the length byte, too.

Uli

**Comment by rafiberlin** [comment]
> yRight 2009 Ver2.4.0316

Hi Uli, thanks for your quick reply!

You mean I should open the GME file with an HEX Editor and arrange the header?

I tried to include in the yaml

 gme-lang: FRENCH
 comment: CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316

When I export the yaml on the created GME with 

_tttool export MyFile.gme_

I can find this in  the Yaml:

comment: CHOMPTECH DATA FORMAT CopyRight 2009 Ver2.4.0316

But I do not see the gme-lang tag. Instead I see 

Language: null 

(I tried to create the GME with _Language: FRENCH_ but it did not work )

**Comment by uli42** [comment]
On Tue, Sep 18, 2018 at 11:36 AM rafiberlin <notifications@github.com> wrote:
>
> yRight 2009 Ver2.4.0316
>
> Hi Uli, thanks for your quick reply!
>
> You mean I should open the GME file with an HEX Editor and arrange the header?

Well, if you use tttool the value should be correct. If you manipulate
with a hex editor you should als adapt the length byte, yes.

Uli

**Comment by nomeata** [comment]
This is old, I hope I can close it. Also note the new support for `tttool set-language`.

**Comment by adomingues** [comment]
Hi @nomeata, 

do happen to know which languages are supported by Tiptoi books? I tried to find out but couldn't find it. 

The reason to ask this, is that I tried to change the audio file for the book "Welt Atlas" to Portuguese with:

```bash
tttool-1.9/tttool set-language PORTUG WWW\ Weltatlas.gme
```

But when I tried the pen with this audio file I got the message "Audio file not found" (or something along these lines in German). I then did a test with "FRENCH" that should work but I got the same message. The original German audio, downloaded from the Tiptoi website, it works. 

Cheers and thank you for developing this tool.
